File: src/com/basmilius/ps/bastools/util/CommandUtils.java
package com.basmilius.ps.bastools.util;

import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class CommandUtils
{

	/**
	 * Runs a command.
	 *
	 * @param project Project that the command should be executed on.
	 * @param command Command that should be executed.
	 * @param handler Handler.
	 */
	public synchronized static void run (final Project project, final String command, final ICommandResultHandler handler)
	{
		if (project == null)
			return;

		ProgressManager.getInstance().run(new Task.Backgroundable(project, "Executing command '" + command + "'")
		{

			@Override
			public void run (@NotNull final ProgressIndicator progress)
			{
				progress.setIndeterminate(true);

				final Runtime runtime = Runtime.getRuntime();

				try
				{
					final Process process = runtime.exec(command, null, new File(project.getBasePath()));
					final BufferedInputStream in = new BufferedInputStream(process.getInputStream());
					final BufferedReader reader = new BufferedReader(new InputStreamReader(in));

					String line;
					final List<String> lines = new ArrayList<>();

					while ((line = reader.readLine()) != null)
					{
						lines.add(line);
					}

					handler.onCommandResult(lines.toArray(new String[lines.size()]));

					process.waitFor();
					reader.close();
					in.close();
				}
				catch (Exception e)
				{
					e.printStackTrace();
				}
			}

		});
	}

	public interface ICommandResultHandler
	{

		void onCommandResult (final String[] lines);

	}

}

File: src/com/basmilius/ps/bastools/util/CommandUtils.kt
package com.basmilius.ps.bastools.util

import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.Project

import java.io.BufferedInputStream
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.util.ArrayList

/**
 * Object CommandUtils
 *
 * @author Bas Milius
 * @package com.basmilius.ps.bastools.util
 */
object CommandUtils
{

	/**
	 * Runs a command.
	 *
	 * @param project Project that the command should be executed on.
	 * @param command Command that should be executed.
	 * @param handler Handler.
	 *
	 * @author Bas Milius
	 */
	@Synchronized
	fun run(project : Project?, command : String, handler : ICommandResultHandler)
	{
		if (project == null)
			return

		ProgressManager.getInstance().run(object : Task.Backgroundable(project, "Executing command '$command'")
		{

			override fun run(progress : ProgressIndicator)
			{
				progress.isIndeterminate = true

				val runtime = Runtime.getRuntime()

				try
				{
					val process = runtime.exec(command, null, File(project.basePath!!))
					val `in` = BufferedInputStream(process.inputStream)
					val reader = BufferedReader(InputStreamReader(`in`))

					var line = reader.readLine()
					val lines = ArrayList<String>()

					while (line != null)
					{
						lines.add(line)
						line = reader.readLine()
					}

					handler.onCommandResult(lines.toTypedArray())

					process.waitFor()
					reader.close()
					`in`.close()
				}
				catch (e : Exception)
				{
					e.printStackTrace()
				}

			}

		})
	}

	/**
	 * Interface ICommandResultHandler
	 *
	 * @author Bas Milius
	 */
	interface ICommandResultHandler
	{

		/**
		 * Invoked when a command is executed.
		 *
		 * @param lines Array<String>
		 *
		 * @author Bas Milius
		 */
		fun onCommandResult(lines : Array<String>)

	}

}